### ISSUE TITLE
Conditional functions with constant expression as condition should evaluate only one branch

### ISSUE THREAD

**Issue by vpanfilov** [ClickHouse/ClickHouse#32969]
For example, if I use UInt8 constant for if condition then only one branch is evaluated:
```sql
SELECT if(1, 'success', throwIf(1, 'Executing FALSE branch'))

┌─'success'─┐
│ success   │
└───────────┘

1 rows in set. Elapsed: 0.003 sec.
```

But if I use some constant expression, then both branches are evaluated:
```sql
SELECT if(empty(''), 'success', throwIf(1, 'Executing FALSE branch'))

0 rows in set. Elapsed: 0.003 sec.

Received exception from server (version 21.12.2):
Code: 395. DB::Exception: Received from localhost:9000. DB::Exception: Executing FALSE branch: While processing if(empty(''), 'success', throwIf(1, 'Executing FALSE branch')). (FUNCTION_THROW_IF_VALUE_IS_NON_ZERO)
```

**Comment by den-crane** [comment]
hm,

```sql
SELECT if(empty(''), 1, sleep(1));
Elapsed: 0.001 sec.

SELECT if(empty(''), 1, toDecimal64(1,2)/0);
DB::Exception: Division by zero:


SELECT if(x = '', 1, toDecimal64(1, 2) / 0)
FROM
(
    SELECT materialize('') AS x
)
DB::Exception: Division by zero:
```

**Comment by kitaisreal** [comment]
Cannot be implemented until #23194.
